core/remote_state_verifier.go

```diff
@@ -0,0 +1,399 @@
+				case types.StatusFullVerified:
+					vt.compareRootHashAndWrite(msg, verifyCh)
+					newRecievedMsgTypeGauge("fullVerified", msg.peerId).Inc(1)
+				case types.StatusPartiallyVerified:
+					log.Warn("block %s , num= %s is untrusted verified", msg.verifyResult.BlockHash, msg.verifyResult.BlockNumber)
+					if vt.allowUntrusted {
+						vt.compareRootHashAndWrite(msg, verifyCh)
+					}
+					newRecievedMsgTypeGauge("partialVerified", msg.peerId).Inc(1)
+				}
+
+			case 2, 4:
+				vt.BadPeers[msg.peerId] = struct{}{}
+				log.Info("peer %s is not available: code %d, msg %s,", msg.peerId, msg.verifyResult.Status.Code, msg.verifyResult.Status.Msg)
+				switch msg.verifyResult.Status {
+				case types.StatusDiffHashMismatch:
+					newRecievedMsgTypeGauge("diffHashMismatch", msg.peerId).Inc(1)
+				case types.StatusImpossibleFork:
+					newRecievedMsgTypeGauge("impossibleFork", msg.peerId).Inc(1)
+				case types.StatusUnexpectedError:
+					newRecievedMsgTypeGauge("unexpectedError", msg.peerId).Inc(1)
+				}
+			case 3:
+				log.Info("return msg from peer %s for block %s is %s", msg.peerId, msg.verifyResult.BlockHash, msg.verifyResult.Status.Msg)
+				switch msg.verifyResult.Status {
+				case types.StatusBlockTooNew:
+					newRecievedMsgTypeGauge("blockTooNew", msg.peerId).Inc(1)
+				case types.StatusBlockNewer:
+					newRecievedMsgTypeGauge("blockNewer", msg.peerId).Inc(1)
+				case types.StatusPossibleFork:
+					newRecievedMsgTypeGauge("possibleFork", msg.peerId).Inc(1)
+				}
+			}
+		case <-resend.C:
+			// if a task has run over 15s, try all the vaild peers to verify.
+			if time.Since(vt.startAt) < tryAllPeersTime {
+				vt.sendVerifyRequest(1)
+			} else {
+				vt.sendVerifyRequest(-1)
+			}
+		case <-vt.terminalCh:
+			return
+		}
+	}
+}
+
+// sendVerifyRequest func select at most n peers from (candidatePeers-badPeers) randomly and send verify request.
+// when n<0, send to all the peers exclude badPeers.
+func (vt *verifyTask) sendVerifyRequest(n int) {
+	var validPeers []VerifyPeer
+	candidatePeers := vt.candidatePeers.GetVerifyPeers()
+	for _, p := range candidatePeers {
+		if _, ok := vt.BadPeers[p.ID()]; !ok {
+			validPeers = append(validPeers, p)
+		}
+	}
+	// if has not valid peer, log warning.
+	if len(validPeers) == 0 {
+		log.Warn("there is no valid peer for block", vt.blockHeader.Number)
+	}
+	if n < 0 || n >= len(validPeers) {
+		for _, p := range validPeers {
+			p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
+		}
+		return
+	}
+
+	// if n < len(validPeers), select n peers from validPeers randomly.
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(validPeers), func(i, j int) { validPeers[i], validPeers[j] = validPeers[j], validPeers[i] })
+	for i := 0; i < n; i++ {
+		p := validPeers[i]
+		p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
+	}
+}
+
+func (vt *verifyTask) compareRootHashAndWrite(msg verifyMessage, verifyCh chan common.Hash) {
```

Comment:
There is no `write` here, please change `compareRootHashAndWrite` to `compareRootHashAndMark`